### ISSUE TITLE
M1, macOS Monterrey 12.1, Vulkan 1.2.198, CMake - missing VK_EXT_metal_surface

### ISSUE THREAD

**Issue by qhdwight** [glfw/glfw#2014]
GLFW v3.3.6 is checkout out as a Git submodule add added via add_subdirectory

CMake:

```cmake
cmake_minimum_required(VERSION 3.16)
project(game_engine)
set(CMAKE_CXX_STANDARD 20)

# GLSLANG
add_subdirectory(glslang)
include_directories(glslang)
# Vulkan
find_package(Vulkan REQUIRED)
include_directories(${Vulkan_INCLUDE_DIRS})
# GLFW
add_subdirectory(glfw)
# GLM
add_subdirectory(glm)
include_directories(glm)
# ENTT
add_subdirectory(entt)
include_directories(entt/src)

file(GLOB SOURCE_FILES CONFIGURE_DEPENDS "src/*.cpp")
add_executable(game_engine ${SOURCE_FILES})
target_link_libraries(game_engine glfw glm glslang glslang-default-resource-limits SPIRV ${PHYSX_LIBRARIES})
target_compile_definitions(game_engine PUBLIC VULKAN_HPP_DISPATCH_LOADER_DYNAMIC=1)
if (WIN32)
    target_compile_definitions(game_engine PUBLIC NOMINMAX VK_USE_PLATFORM_WIN32_KHR)
elseif(APPLE)
    target_compile_definitions(game_engine PUBLIC VK_USE_PLATFORM_MACOS_MVK)
endif ()
```

I get the error:

```
glfw: (65542) Cocoa: Vulkan instance missing VK_EXT_metal_surface extension
```

I have installed the official Vulkan SDK for macOS and linked all the system files. The context creation is handled by [this](https://github.com/KhronosGroup/Vulkan-Hpp/blob/master/samples/utils/utils.cpp) file. I have made sure that "VK_MVK_macos_surface" is added to the init extensions by means of the debugger.

Please let me know what other details help!

**Comment by ByunghoonKim** [comment]
Your "game_engine" expects `VK_EXT_metal_surface` extension, but you gave it `VK_MVK_macos_surface` instead. Both are valid Vulkan instance extensions for macOS, but the former is a newer addition. Either giving what your engine expects, or changing its expectation would solve the issue.

**Comment by elmindreda** [comment]
Call [glfwGetRequiredInstanceExtensions](https://www.glfw.org/docs/latest/vulkan_guide.html#vulkan_ext) to find out which extensions GLFW requires for window surface creation.

**Comment by ByunghoonKim** [comment]
@qhdwight, I hope you've already figured it out, but here is more detailed explanation on what happens in your code, and a quick & dirty fix for you. Just in case, because it's a holiday. :)

In your CMake code, this
```CMake
elseif(APPLE)
    target_compile_definitions(game_engine PUBLIC VK_USE_PLATFORM_MACOS_MVK)
endif ()
```
triggers line 561-562 of [util.h](https://github.com/KhronosGroup/Vulkan-Hpp/blob/master/samples/utils/utils.cpp) you linked,
```C
#elif defined( VK_USE_PLATFORM_MACOS_MVK )
      extensions.push_back( VK_MVK_MACOS_SURFACE_EXTENSION_NAME );
```
which requests the Vulkan instance extension `VK_MVK_macos_surface` to be enabled. Your error message, however,
```
glfw: (65542) Cocoa: Vulkan instance missing VK_EXT_metal_surface extension
```
tells, when your game_engine tries to create a Vulkan instance, it expects `VK_EXT_metal_surface` extension enabled instead. Hence, for the time being, simply replacing, in util.h, 
```C
#elif defined( VK_USE_PLATFORM_MACOS_MVK )
      extensions.push_back( VK_MVK_MACOS_SURFACE_EXTENSION_NAME );
```
with
```C
#elif defined( VK_USE_PLATFORM_MACOS_MVK )
      //extensions.push_back( VK_MVK_MACOS_SURFACE_EXTENSION_NAME );
      extensions.push_back( VK_EXT_METAL_SURFACE_EXTENSION_NAME );
```
should solve your problem. And because `VK_MVK_macos_surface` is deprecated anyway, you're not losing anything here.

**Comment by ByunghoonKim** [comment]
@elmindreda In Vulkan support reference page, [glfwCreateWindowSurface](https://www.glfw.org/docs/latest/group__vulkan.html#ga1a24536bec3f80b08ead18e28e6ae965) says

> macOS: This function currently only supports the VK_MVK_macos_surface extension from MoltenVK.

I think I slipped to fix that line to include `VK_EXT_metal_surface` when I first submitted the pull request(#1619).

**Comment by qhdwight** [comment]
Thank you for the replies all. Funnily enough I just happened to get it working, and was checking back to post my findings before seeings your solution. I did have to do some additional things forever:

In `utils.cpp` `getInstanceExtensions`:
```c++
extensions.push_back( "VK_EXT_metal_surface" );
extensions.push_back( "VK_KHR_get_physical_device_properties2" );
```

In `utils.cpp` `getDeviceExtensions`:
```c++
return { VK_KHR_SWAPCHAIN_EXTENSION_NAME, "VK_KHR_portability_subset" };
```

I saw that info about the other extensions [here](https://stackoverflow.com/questions/66659907/vulkan-validation-warning-catch-22-about-vk-khr-portability-subset-on-moltenvk). Not sure why it is really needed as of right now.

The `glfwGetRequiredInstanceExtensions` note is helpful, but it's awkward to integrate with the Khronos group code. Maybe I'll transition to my own boilerplate...

I may get around to making a pull request about the deprecated macOS surface type over on [their repo](https://github.com/KhronosGroup/Vulkan-Hpp). Doesn't seem like a GLFW issue at all. But maybe some changes in documentation could help nudge people in the right direction.
